source/base/StringBuilder.ooc

```diff
@@ -0,0 +1,64 @@
+/*
+* Copyright (C) 2015 - Simon Mika <simon@mika.se>
+*
+* This sofware is free software; you can redistribute it and/or
+* modify it under the terms of the GNU Lesser General Public
+* License as published by the Free Software Foundation; either
+* version 2.1 of the License, or (at your option) any later version.
+*
+* This software is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	See the GNU
+* Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public License
+* along with this software. If not, see <http://www.gnu.org/licenses/>.
+*/
+use ooc-collections
+
+StringBuilder: class {
+	_stringList := VectorList<String> new()
+
+	count ::= this _stringList count
+
+	init: func ~default
+	init: func ~string (value: String) {
+		this init()
+		this _stringList add(value)
+	}
+	init: func ~this (orig: This) {
+		this init()
+		orig _stringList apply( func (value: String) { this _stringList add(value) })
+	}
+	copy: func -> This {
+		This new(this)
+	}
+
+	add: func (value: String) {
+		this _stringList add(value)
+	}
```

Comment:
So, `Strings` are added to the list. When the `StringBuilder` is freed, it should also free its list. But what happens to the contents in the list? Should they be freed? If you want to keep the `String` you add to the `StringBuilder`, should you give the `StringBuilder` a copy of the `String`?

```ooc
sb := StringBuilder new()
myString := "Hello World"
sb add(myString)
sb toString() println().free()
sb free() // the list in s is also free, and so are all the strings in the list
myString println() // this is going to crash
```
so should we instead do
```ooc
sb := StringBuilder new()
myString := "Hello World"
sb add(myString copy())
sb toString() println().free() // free the String created by the `toString()` call
sb free() // the list in s is also freed, and so are all the Strings in the list
myString println() // this is should work
```
? Having to type `copy()` in each `add()` call is going to get messy. If we modify `add()` like this
```ooc
	add: func (value: String) {
		this _stringList add(value copy())
	}
```
we don't have to type
```ooc
sb add(myString copy())
```
but can do
```ooc
sb add(myString)
```
, so that's nice. But what if we don't want to keep `myString`? Then we'll have to do
```ooc
sb add(myString)
myString free()
```
and that's going to get messy, too. In addition, we copy a `String` and then free the original immediately afterwards, which is unnecessary work.

One way to fix this is by adding a second argument to `add`:
```ooc
	add: func (value: String, copy := true) {
		this _stringList add(value copy())
	}
```
but
```ooc
sb add(myString, false)
```
is also not very pretty and people are likely to make mistakes.

Another solution is to give `StringBuilder` operators for appending and freeing (or not) the argument, but then we haven't gained anything by adding `StringBuilder`, because the whole point is that the person using it shouldn't have to worry about memory leaks, or else they're just as well off using `String` operators directly.